### ISSUE TITLE
Animate to and rotate around a point in lower half of the screen

### ISSUE THREAD

**Issue by pasniak** [osmdroid/osmdroid#1294]
## Issue Type

[x] Question
[ ] Bug
[x] Improvement
[ ] Build system related
[ ] Performance
[ ] Documentation

I am trying animateTo to go to a point roughly 2/3 down the screen (rather than in the middle) to show more space in front of when map is rotated to the user's heading. It seems I need to override MapController but it there is no setter. Or maybe I am missing some obvious solution?

**Comment by pasniak** [comment]
I found `setExpectedCenter(lastLocation.toGeoPoint(), 0, 300)` which (kind of) shifts the center, yet there is not equivalent `animateTo` (which looks more neat). 

When I set `mapOrientation` <> 0 (rotate the map to compass) the point is shifted 300 points but not down the screen... and the rotation happens around the "middle of the screen" and not the "center" due to /2 hardcoded in:
```
	private void invalidateMapCoordinates(int left, int top, int right, int bottom, boolean post) {
		mInvalidateRect.set(left, top, right, bottom);

		final int centerX = getWidth() / 2;
		final int centerY = getHeight() / 2;

		if (this.getMapOrientation() != 0)
			GeometryMath.getBoundingBoxForRotatatedRectangle(mInvalidateRect, centerX, centerY,
					this.getMapOrientation() + 180, mInvalidateRect);
```

**Comment by pasniak** [comment]
Clarification. The effect I am after is: the map (rotated to direction of vehicle's movement) would show more space in front of the vehicle.

**Comment by spyhunter99** [comment]
there was an open branch on a feature related to this. Someone had opened a PR but never provide a demo or instructions on how to use it. I never got it to work correctly but maybe someone else can. I think it was called advanced navigation or something to that effect

**Comment by monsieurtanuki** [comment]
@pasniak Let me rephrase your OP.
So far the most comprehensive version of `MapController.animateTo` has the following parameters, and its reference is the center of the `MapView`:
````java
    final IGeoPoint point, final Double pZoom, final Long pSpeed,
    final Float pOrientation, final Boolean pClockwise
````

You'd like to add the following parameters:
````java
    final int pOffsetX,
    final int pOffsetY
````
And have it work exactly the same, _except_ that a pixel offset is applied to the `MapView`.

Working on it. It should be quick.

**Comment by monsieurtanuki** [comment]
@pasniak Just PR'ed #1306.
Works OK with an orientation of 0, but I'm not sure if it suits your needs when the map is oriented.

**Comment by pasniak** [comment]
The point of this humble request is for display work when map rotates to compass or direction of movemement (so more space ahead - in movement's direction - is visible).

**Comment by monsieurtanuki** [comment]
@pasniak I'm not sure I can visualize exactly what you want.
What I did in the new "More Samples / Events / Animate To with Offsets" is something like that, with a (0, 200) pixel offset:
![20190402 offset0_200 rotation0](https://user-images.githubusercontent.com/11576431/55384620-79fba380-552b-11e9-9b1f-9c2bc5457da0.png)
...with an orientation of 0, and the next one is with an orientation of -30°:
![20190402 offset0_200 rotation-30](https://user-images.githubusercontent.com/11576431/55384673-9566ae80-552b-11e9-9a16-e43b8ae8ab85.png)

The red line goes from the center of the screen (previous animateTo target) to the GeoPoint we're animating to.
Do those screenshots suit you? Would you prefer the red line to have always the same orientation, meaning basically that regardless of the map orientation, the GeoPoint animated to will always be at the same pixel position on the screen?

**Comment by pasniak** [comment]
>Would you prefer the red line to have always the same orientation, meaning basically that regardless of the map orientation, the GeoPoint animated to will always be at the same pixel position on the screen?

Exactly yes!

**Comment by monsieurtanuki** [comment]
@pasniak Something like that, then:
![Screenshot_2019-04-04-16-02-29](https://user-images.githubusercontent.com/11576431/55561920-466b7580-56f3-11e9-91f0-7213d42481b1.png)

**Comment by monsieurtanuki** [comment]
@pasniak Feel free to test my latest PR.

**Comment by pasniak** [comment]
I played with animateTo with offset but I got close to the desired effect is by _only_ overriding `setExpectedCenter`:
```
    override fun setExpectedCenter(pGeoPoint: IGeoPoint?) { // not needed?
        super.setExpectedCenter(pGeoPoint, 0, 500)
    }

    override fun setExpectedCenter(pGeoPoint: IGeoPoint?, pOffsetX: Long, pOffsetY: Long) {
        super.setExpectedCenter(pGeoPoint, pOffsetX, 500)
    }
```

Maybe I am doing sth wrong but my animateTos are strange...:
![Peek 2019-04-05 07-47](https://user-images.githubusercontent.com/19778666/55625636-e62f0f00-5798-11e9-8e0a-0d4d93656df6.gif)
Debugging it is a bit tough. At first the maps seems scrolled twice offset then  the position is offset the right amount 500 or sth like that :)

I call animateTo from few locations in the app and overriding setExpectedCenter seemed today like good central place to try setting the map's center offset.

**Comment by monsieurtanuki** [comment]
@pasniak Can we at least agree on the fact that the behavior of the "More Samples / Events / Animate To with Offsets" demo matches the expectations?

**Comment by pasniak** [comment]
The reason I tried the above was the "Animate To with Offsets" works weirdly when I changed it to set map orientation the heading between the points; it does not move in straight lines but in curved circles between points....

**Comment by monsieurtanuki** [comment]
I can explain the curve effect: in order to take into account the map orientation (when not 0), I apply a mix of the offset angle (in a vertical case 90 or -90) and the orientation, with `cos` and `sin`.

Besides the curve effect during the animation, is the `Marker` correctly positioned at the end of the animation?

**Comment by pasniak** [comment]
Yes, the position is correct. 

The curves are appearing when:
```
        final Float orientation = mIndex>0 ? (float) mList.get((mIndex - 1)%mList.size()).bearingTo(mList.get(mIndex%mList.size())) : null;
```

**Comment by pasniak** [comment]
Maybe in the beginning I got fixated on `animateTo` because the positioning is there? 

IMHO the offset seems map feature independent from position, just like orientation and better to set it along `mapView.setMapOrientation`, basically I'd want to set offset once and with every `animateTo` call. At least this is my use case (other use case where you'd want offset in `animateTo` is maybe leave space for the label)

**Comment by monsieurtanuki** [comment]
If I understand well, it would be something like `mapView.setMapCenterOffset(int pixelX, int pixelY)`:
* the normal behavior of `mapView.setExpectedCenter(GeoPoint)` is to make that `GeoPoint` match the center of the screen
* with the offsets, we can make it match any other pixel of the screen

Is that what you have in mind?
Btw I won't be able to work on it before June, next week I'm on the Way of Saint-James.

**Comment by pasniak** [comment]
That is allright.  Here's how I would add it to my code:
```
        private val gpsListener = rxBusSubscribe<GPSEvent> {
...
                mapViewWrap.setOrientation(this.loc.bearing)
                mapViewWrap.setMapCenterOffset(0, 500) // most likely 1/3 of screen height or speed dependant
            }
        }
```

**Comment by monsieurtanuki** [comment]
@pasniak Just PR'ed #1353 - feel free to test it.

**Comment by monsieurtanuki** [comment]
Fixed by #1353
